### PR TITLE
[NativeAOT-LLVM] Merge to August 6

### PR DESCRIPTION
**Pull request by SingleAccretion**
Notable changes:
1) WASI SDK version update to 24 and the switch to WASI P2 (following upstream). Includes using wasmtime for testing.
2) Emscripten update to 3.1.56 (following upstream).
3) Adjust the warning check for old WASI SDKs to work around https://github.com/WebAssembly/wasi-sdk/issues/471.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/nativeaot-llvm

File: eng/native/gen-buildsys.sh [hunk 1]
```diff
@@ -99,16 +99,7 @@ if [[ "$host_arch" == "wasm" ]]; then
     if [[ "$target_os" == "browser" ]]; then
         cmake_command="emcmake $cmake_command"
     elif [[ "$target_os" == "wasi" ]]; then
-<<<<<<< HEAD
-        if [[ -z $WASI_SDK_PATH ]]; then
```
Comment (yowl) [inline eng/native/gen-buildsys.sh:101]: This check has no value anymore?

**Review by yowl — COMMENTED** [review]
Just looked at the resolve conflict commit so far.

File: eng/pipelines/runtimelab/install-wasi-sdk.ps1 [hunk 2]
```diff
@@ -1,13 +1,20 @@
-Invoke-WebRequest -Uri https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-22/wasi-sdk-22.0.m-mingw64.tar.gz -OutFile wasi-sdk-22.0.m-mingw64.tar.gz
+$WasiSdkVersion = 24
 
-tar -xzf wasi-sdk-22.0.m-mingw64.tar.gz
+# Verify that we're not behind upstream (and allow us to be ahead).
+$UpstreamWasiSdkVersion = Get-Content $PSScriptRoot/../../../src/mono/wasi/wasi-sdk-version.txt
+if ($WasiSdkVersion -lt [int]$UpstreamWasiSdkVersion)
+{
+    Write-Error "Upstream WASI SDK version is $UpstreamWasiSdkVersion; update `$WasiSdlVersion (currently $WasiSdlVersion)!"
```
Comment (yowl) [inline eng/pipelines/runtimelab/install-wasi-sdk.ps1]:
```suggestion
    Write-Error "Upstream WASI SDK version is $UpstreamWasiSdkVersion; update `$WasiSdlVersion (currently $WasiSdkVersion)!"
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets [hunk 3]
```diff
@@ -545,10 +551,12 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeLibrary Condition="'$(IlcLlvmExceptionHandlingModel)' == 'cpp'" Include="$(IlcSdkPath)libCppExceptionHandling.a" />
       <NativeLibrary Condition="'$(IlcLlvmExceptionHandlingModel)' == 'wasm'" Include="$(IlcSdkPath)libWasmExceptionHandling.a" />
       <NativeLibrary Condition="'$(IlcLlvmExceptionHandlingModel)' == 'emulated'" Include="$(IlcSdkPath)libEmulatedExceptionHandling.a" />
+      <_WASIComponentImports Include="$(IlcFrameworkNativePath)*.wit" />
```
Comment (yowl) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Are we ok mixing all caps, `WASI` and pascal, e.g. `Ilc` ?

**Review by yowl — COMMENTED** [review]
minor comments

**Review by yowl — APPROVED** [review]
Thanks!, minor comments is all.

File: eng/native/gen-buildsys.sh (hunk 1 above)
Comment (SingleAccretion) [reply eng/native/gen-buildsys.sh:101]: It's more that it isn't in the upstream file (it is a goal to have little divergence). I will bring it back.

File: eng/pipelines/runtimelab/install-wasi-sdk.ps1 (hunk 2 above)
Comment (SingleAccretion) [reply eng/pipelines/runtimelab/install-wasi-sdk.ps1]: Fixed.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 3 above)
Comment (SingleAccretion) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Changed to `Wasi`. It's intentionally similar to the item name used in `wasi.proj`.

**Review by pavelsavara — APPROVED** [review]

**Comment by hez2010** [comment]
Can you seek the merge to the latest commit (particularly, after [99f7f934927c7e2c2af2491cb9b9aa2063a562d2](https://github.com/dotnet/runtime/commit/99f7f934927c7e2c2af2491cb9b9aa2063a562d2))? I would like to test the Avalonia wasm size saving after it.

**Comment by jkotas** [comment]
Thank you!

**Comment by SingleAccretion** [comment]
> Can you seek the merge to the latest commit (particularly, after [99f7f934927c7e2c2af2491cb9b9aa2063a562d2](https://github.com/dotnet/runtime/commit/99f7f934927c7e2c2af2491cb9b9aa2063a562d2))? I would like to test the Avalonia wasm size saving after it.

@hez2010 Historically, we've preferred not to do merges too frequently (they are disruptive). An exception can always be made: https://github.com/dotnet/runtimelab/pull/2653.

**Pull request merged**